src/d3d8/d3d8_device.cpp

```diff
@@ -994,29 +994,46 @@ namespace dxvk {
     HRESULT res = GetD3D9()->CreateStateBlock(d3d9::D3DSTATEBLOCKTYPE(Type), &pStateBlock9);
 
     D3D8StateBlock* pStateBlock = new D3D8StateBlock(this, Type, pStateBlock9.ref());
-    m_stateBlocks.insert(pStateBlock);
 
-    *pToken = DWORD(reinterpret_cast<uintptr_t>(pStateBlock));
+    m_stateBlocks.insert({m_token, pStateBlock});
+    *pToken = m_token;
+    m_token++;
 
     return res;
   }
 
   HRESULT STDMETHODCALLTYPE D3D8Device::CaptureStateBlock(DWORD Token) {
-    return reinterpret_cast<D3D8StateBlock*>(Token)->Capture();
+    if (unlikely(m_stateBlocks.find(Token) == m_stateBlocks.end())) {
+      Logger::err("Invalid token passed to CaptureStateBlock");
+      return D3DERR_INVALIDCALL;
+    }
+
+    return m_stateBlocks.at(Token)->Capture();
   }
 
   HRESULT STDMETHODCALLTYPE D3D8Device::ApplyStateBlock(DWORD Token) {
     StateChange();
-    return reinterpret_cast<D3D8StateBlock*>(Token)->Apply();
+
+    if (unlikely(m_stateBlocks.find(Token) == m_stateBlocks.end())) {
+      Logger::err("Invalid token passed to ApplyStateBlock");
+      return D3DERR_INVALIDCALL;
+    }
+
+    return m_stateBlocks.at(Token)->Apply();
   }
 
   HRESULT STDMETHODCALLTYPE D3D8Device::DeleteStateBlock(DWORD Token) {
     // "Applications cannot delete a device-state block while another is being recorded"
     if (unlikely(ShouldRecord()))
       return D3DERR_INVALIDCALL;
 
-    D3D8StateBlock* block = reinterpret_cast<D3D8StateBlock*>(Token);
-    m_stateBlocks.erase(block);
+    if (unlikely(m_stateBlocks.find(Token) == m_stateBlocks.end())) {
```

Comment:
I said before in my review to save off the iterator from find. Right now you are looking it up twice for no reason.

Reply:
Hopefully fine now :sweat_smile:.